formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt

```diff
@@ -467,3 +481,74 @@ private fun Iterable<ByteArray>.flatten(): ByteArray {
 private fun SerialDescriptor.isByteString(index: Int): Boolean =
     getElementDescriptor(index) == ByteArraySerializer().descriptor &&
         getElementAnnotations(index).find { it is ByteString } != null
+
+private fun halfToDouble(bits: Short): Long {
+    val sign = (bits.toLong() shr 15 and 0x0001)
+    val halfExp = bits.toLong() shr 10 and 0x1f //max half exponent value
+    val halfMant = bits.toLong() and 0x3ff //max half mantis value
+
+    var exp: Long
+    var mant: Long
+
+    if (halfExp == 0x1fL) {//max half exponent value
+        //if mantissa maximal - value is NaN or Infinity
+        exp = 0x7FF //max double exponent value
+        mant = halfMant
+    } else if (halfExp == 0L) {
+        if (halfMant == 0L) {
+            //if exponent and mantissa are zero - value is zero
+            mant = 0
+            exp = 0
+        } else {
+            //if exponent is zero and mantissa non-zero - value denormalized. normalize it
+            exp = halfExp + 1008 + 1 //double exponent bias - half exponent bias = 1023 - 15
+            mant = halfMant
+            while (mant <= 0x3ff) {//max half mantis value
+                mant = mant shl 1
+                exp--
+            }
+
+            mant = mant and 0x3ff //max half mantis value
+        }
+    } else {
+        //normalized value
+        exp = (halfExp + 1008) //double exponent bias - half exponent bias = 1023 - 15
+        mant = halfMant
+    }
+    return (sign shl 63) or (exp shl 52) or (mant shl 42)
+}
```

Comment:
The [example code provided in the CBOR RFC](https://tools.ietf.org/html/rfc7049#appendix-D) seems a bit simpler and doesn't have a loop. Are there specific advantages to this algorithm as apposed to the one listed in the RFC?

Reply:
Not sure RFC's one work with denormalized numbers. Btw, sources of AOSP (https://android.googlesource.com/platform/libcore/+/master/luni/src/main/java/libcore/util/FP16.java#578) are using magic constant instead of a loop.

Reply:
Examples in RFC use function ldexp(man, pow) which returns  (man * 2^pow). There is no same function in Kotlin so alternative is to work with bits by yourself.

Reply:
"(https://android.googlesource.com/platform/libcore/+/master/luni/src/main/java/libcore/util/FP16.java#578" - it is a good trick with 0.5 float value!
FPU normalize value faster then shifts in a loop.